File: backend/routers/career.py
import datetime as dt
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.orm import Session

from backend.deps import get_db, rows_to_list, row_to_dict
from db.models import Employment, Education, Certification, Project, Achievement

router = APIRouter(tags=["career"])


class EmploymentIn(BaseModel):
    organization: str
    role: str
    start_date: Optional[dt.date] = None
    end_date: Optional[dt.date] = None
    is_current: bool = False


class EducationIn(BaseModel):
    school: str
    degree: str = ""
    field: str = ""
    start_date: Optional[dt.date] = None
    end_date: Optional[dt.date] = None


class CertificationIn(BaseModel):
    name: str
    issuer: str = ""
    date_earned: Optional[dt.date] = None
    expires: Optional[dt.date] = None


@router.get("/employments")
def list_employments(db: Session = Depends(get_db)):
    return rows_to_list(db.execute(select(Employment).order_by(Employment.start_date.desc())).scalars().all())


@router.post("/employments")
def create_employment(payload: EmploymentIn, db: Session = Depends(get_db)):
    e = Employment(**payload.model_dump())
    db.add(e)
    db.commit()
    return row_to_dict(e)


@router.delete("/employments/{employment_id}")
def delete_employment(employment_id: int, db: Session = Depends(get_db)):
    e = db.get(Employment, employment_id)
    if not e:
        raise HTTPException(404, "Not found")
    db.delete(e)
    db.commit()
    return {"ok": True}


@router.get("/education")
def list_education(db: Session = Depends(get_db)):
    return rows_to_list(db.execute(select(Education)).scalars().all())


@router.post("/education")
def create_education(payload: EducationIn, db: Session = Depends(get_db)):
    e = Education(**payload.model_dump())
    db.add(e)
    db.commit()
    return row_to_dict(e)


@router.delete("/education/{education_id}")
def delete_education(education_id: int, db: Session = Depends(get_db)):
    e = db.get(Education, education_id)
    if not e:
        raise HTTPException(404, "Not found")
    db.delete(e)
    db.commit()
    return {"ok": True}


@router.get("/certifications")
def list_certifications(db: Session = Depends(get_db)):
    return rows_to_list(db.execute(select(Certification)).scalars().all())


@router.post("/certifications")
def create_certification(payload: CertificationIn, db: Session = Depends(get_db)):
    c = Certification(**payload.model_dump())
    db.add(c)
    db.commit()
    return row_to_dict(c)


@router.delete("/certifications/{cert_id}")
def delete_certification(cert_id: int, db: Session = Depends(get_db)):
    c = db.get(Certification, cert_id)
    if not c:
        raise HTTPException(404, "Not found")
    db.delete(c)
    db.commit()
    return {"ok": True}


@router.get("/timeline")
def timeline(db: Session = Depends(get_db)):
    events = []
    for e in db.execute(select(Employment)).scalars().all():
        if e.start_date:
            events.append({"date": e.start_date.isoformat(), "label": f"Started {e.role} at {e.organization}"})
        if e.end_date and not e.is_current:
            events.append({"date": e.end_date.isoformat(), "label": f"Ended role at {e.organization}"})
    for p in db.execute(select(Project)).scalars().all():
        if p.start_date:
            events.append({"date": p.start_date.isoformat(), "label": f"Started project {p.name}"})
    for a in db.execute(select(Achievement)).scalars().all():
        events.append({"date": a.created_at.date().isoformat(), "label": f"Achievement: {a.title}"})
    events.sort(key=lambda e: e["date"])
    return events

File: frontend/static/js/pages/career.js
Pages.career = {
  async render(container) {
    container.innerHTML = `
      <div class="page-header"><h1 class="page-title">Career Journey</h1></div>
      <div class="btn-row" id="career-tabs">
        <button class="btn-secondary tab-btn active" data-tab="employment">Employment</button>
        <button class="btn-secondary tab-btn" data-tab="education">Education</button>
        <button class="btn-secondary tab-btn" data-tab="certifications">Certifications</button>
        <button class="btn-secondary tab-btn" data-tab="timeline">Timeline</button>
      </div>
      <div class="section-gap" id="career-panel"></div>
    `;

    const panel = container.querySelector("#career-panel");
    const tabButtons = [...container.querySelectorAll(".tab-btn")];

    const panels = { employment: renderEmployment, education: renderEducation, certifications: renderCertifications, timeline: renderTimeline };

    async function showTab(tab) {
      tabButtons.forEach((b) => b.classList.toggle("active", b.dataset.tab === tab));
      tabButtons.forEach((b) => b.style.background = b.dataset.tab === tab ? "var(--accent-soft)" : "");
      panel.innerHTML = emptyState("Loading…");
      await panels[tab](panel);
    }

    tabButtons.forEach((btn) => btn.addEventListener("click", () => showTab(btn.dataset.tab)));
    showTab("employment");
  },
};

async function renderEmployment(panel) {
  const employments = await Api.employments();
  panel.innerHTML = `
    <details class="expandable">
      <summary>Add employment</summary>
      <form id="emp-form" class="section-gap">
        <div class="field"><label>Organization*</label><input type="text" id="emp-org" required /></div>
        <div class="field"><label>Role*</label><input type="text" id="emp-role" required /></div>
        <div class="form-row">
          <div class="field"><label>Start Date</label><input type="date" id="emp-start" value="${todayStr()}" /></div>
          <div class="field"><label>End Date</label><input type="date" id="emp-end" /></div>
        </div>
        <label class="small"><input type="checkbox" id="emp-current" checked /> Current</label>
        <div class="btn-row"><button type="submit" class="btn-primary">Save</button></div>
      </form>
    </details>
    <div class="card section-gap">
      ${employments.length ? employments.map((e) => `
        <div class="entry-item">
          <div class="entry-item-head">
            <span class="entry-item-title">${escapeHtml(e.organization)} — ${escapeHtml(e.role)}</span>
            <button class="btn-danger" data-del="${e.id}">Delete</button>
          </div>
          <div class="entry-item-body">${formatDate(e.start_date)} – ${e.is_current ? "present" : formatDate(e.end_date)}</div>
        </div>`).join("") : emptyState("No employment history yet.")}
    </div>
  `;
  panel.querySelector("#emp-form").addEventListener("submit", async (e) => {
    e.preventDefault();
    const organization = panel.querySelector("#emp-org").value.trim();
    const role = panel.querySelector("#emp-role").value.trim();
    if (!organization || !role) return;
    await Api.createEmployment({
      organization, role,
      start_date: panel.querySelector("#emp-start").value || null,
      end_date: panel.querySelector("#emp-end").value || null,
      is_current: panel.querySelector("#emp-current").checked,
    });
    toast("Saved");
    renderEmployment(panel);
  });
  panel.querySelectorAll("[data-del]").forEach((btn) => btn.addEventListener("click", async () => {
    if (!confirm("Delete this employment record?")) return;
    await Api.deleteEmployment(btn.dataset.del);
    renderEmployment(panel);
  }));
}

async function renderEducation(panel) {
  const items = await Api.education();
  panel.innerHTML = `
    <details class="expandable">
      <summary>Add education</summary>
      <form id="edu-form" class="section-gap">
        <div class="field"><label>School*</label><input type="text" id="edu-school" required /></div>
        <div class="form-row">
          <div class="field"><label>Degree</label><input type="text" id="edu-degree" /></div>
          <div class="field"><label>Field of study</label><input type="text" id="edu-field" /></div>
        </div>
        <div class="form-row">
          <div class="field"><label>Start Date</label><input type="date" id="edu-start" /></div>
          <div class="field"><label>End Date</label><input type="date" id="edu-end" /></div>
        </div>
        <div class="btn-row"><button type="submit" class="btn-primary">Save</button></div>
      </form>
    </details>
    <div class="card section-gap">
      ${items.length ? items.map((e) => `
        <div class="entry-item">
          <div class="entry-item-head">
            <span class="entry-item-title">${escapeHtml(e.school)}</span>
            <button class="btn-danger" data-del="${e.id}">Delete</button>
          </div>
          <div class="entry-item-body">${escapeHtml(e.degree || "")} ${e.field ? "in " + escapeHtml(e.field) : ""} · ${formatDate(e.start_date)} – ${e.end_date ? formatDate(e.end_date) : "present"}</div>
        </div>`).join("") : emptyState("No education records yet.")}
    </div>
  `;
  panel.querySelector("#edu-form").addEventListener("submit", async (e) => {
    e.preventDefault();
    const school = panel.querySelector("#edu-school").value.trim();
    if (!school) return;
    await Api.createEducation({
      school,
      degree: panel.querySelector("#edu-degree").value,
      field: panel.querySelector("#edu-field").value,
      start_date: panel.querySelector("#edu-start").value || null,
      end_date: panel.querySelector("#edu-end").value || null,
    });
    toast("Saved");
    renderEducation(panel);
  });
  panel.querySelectorAll("[data-del]").forEach((btn) => btn.addEventListener("click", async () => {
    if (!confirm("Delete this education record?")) return;
    await Api.deleteEducation(btn.dataset.del);
    renderEducation(panel);
  }));
}

async function renderCertifications(panel) {
  const items = await Api.certifications();
  panel.innerHTML = `
    <details class="expandable">
      <summary>Add certification</summary>
      <form id="cert-form" class="section-gap">
        <div class="field"><label>Certification name*</label><input type="text" id="cert-name" required /></div>
        <div class="field"><label>Issuer</label><input type="text" id="cert-issuer" /></div>
        <div class="form-row">
          <div class="field"><label>Date earned</label><input type="date" id="cert-earned" value="${todayStr()}" /></div>
          <div class="field"><label>Expires</label><input type="date" id="cert-expires" /></div>
        </div>
        <div class="btn-row"><button type="submit" class="btn-primary">Save</button></div>
      </form>
    </details>
    <div class="card section-gap">
      ${items.length ? items.map((c) => `
        <div class="entry-item">
          <div class="entry-item-head">
            <span class="entry-item-title">${escapeHtml(c.name)}</span>
            <button class="btn-danger" data-del="${c.id}">Delete</button>
          </div>
          <div class="entry-item-body">${escapeHtml(c.issuer || "")} · earned ${formatDate(c.date_earned)}${c.expires ? " · expires " + formatDate(c.expires) : ""}</div>
        </div>`).join("") : emptyState("No certifications yet.")}
    </div>
  `;
  panel.querySelector("#cert-form").addEventListener("submit", async (e) => {
    e.preventDefault();
    const name = panel.querySelector("#cert-name").value.trim();
    if (!name) return;
    await Api.createCertification({
      name,
      issuer: panel.querySelector("#cert-issuer").value,
      date_earned: panel.querySelector("#cert-earned").value || null,
      expires: panel.querySelector("#cert-expires").value || null,
    });
    toast("Saved");
    renderCertifications(panel);
  });
  panel.querySelectorAll("[data-del]").forEach((btn) => btn.addEventListener("click", async () => {
    if (!confirm("Delete this certification?")) return;
    await Api.deleteCertification(btn.dataset.del);
    renderCertifications(panel);
  }));
}

async function renderTimeline(panel) {
  const events = await Api.timeline();
  panel.innerHTML = `
    <div class="card">
      <div class="small muted" style="margin-bottom:12px;">Auto-built from employment, project, and achievement dates.</div>
      ${events.length ? `<div class="evidence-chain">${events.map((e) => `
        <div class="step"><span class="step-label">${formatDate(e.date)}</span><div>${escapeHtml(e.label)}</div></div>
      `).join("")}</div>` : emptyState("No timeline events yet — add employment, projects, or achievements.")}
    </div>
  `;
}
